### ISSUE TITLE
Improvements of DataObject

### ISSUE THREAD

**Issue by wangcj05** [idaholab/raven#839]
--------
Issue Description
--------
##### What did you expect to see happen?
DataObject should fast enough to load and print 

##### What did you see instead?
The computational time will increase very quickly with the increase of number variables: i.e. computational time t ~ [number of variables]^2

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task? __task__
- [x] 2. Is it tagged with a priority: critical, normal or minor? __critical__
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements? __yes, this PR will change one requirement test, see issue #842 __
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. __not a defect__
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) __yes__

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

**Comment by wangcj05** [comment]
profile information
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    28576   49.418    0.002   49.418    0.002 {method 'index' of 'list' objects}
        1   13.851   13.851   55.179   55.179 DataSet.py:919(_convertArrayListToDataset)
  1770881   11.382    0.000   27.710    0.000 collections.py:50(__init__)
24507980/24507976    7.442    0.000   11.484    0.000 {isinstance}
  2570663    5.472    0.000    5.472    0.000 collections.py:71(__setitem__)
  1770912    4.509    0.000   12.942    0.000 _abcoll.py:548(update)
   799850    4.101    0.000    4.101    0.000 {method 'reduce' of 'numpy.ufunc' objects}
6086771/6001087    3.614    0.000    6.669    0.000 {hasattr}
        2    3.342    1.671    4.144    2.072 {method 'read' of 'pandas._libs.parsers.TextReader' objects}
  1199577    3.208    0.000    8.274    0.000 variable.py:140(as_compatible_data)
  1256716    2.924    0.000    4.087    0.000 collections.py:125(items)
  7083409    2.858    0.000    2.858    0.000 collections.py:90(__iter__)
  3513948    2.833    0.000    2.833    0.000 _weakrefset.py:70(__contains__)
2228740/2143055    2.777    0.000    3.066    0.000 {numpy.core.multiarray.array}
       55    2.610    0.047    2.610    0.047 {pandas._libs.lib.write_csv_rows}
142819/142815    2.455    0.000    8.176    0.000 base.py:152(__new__)
13110460/11224184    2.394    0.000    3.522    0.000 {len}
4995946/4995941    2.362    0.000    3.332    0.000 {getattr}
  2856136    2.157    0.000    6.103    0.000 utils.py:425(ndim)
   199927    2.115    0.000   25.143    0.000 duck_array_ops.py:275(f)
   257049    2.078    0.000    6.159    0.000 nanfunctions.py:107(_divide_by_count)
   142810    2.051    0.000   18.462    0.000 merge.py:110(merge_variables)
   999649    2.005    0.000   15.541    0.000 variable.py:242(__init__)
  1171508    1.966    0.000    3.413    0.000 abc.py:128(__instancecheck__)
  3741543    1.935    0.000    4.439    0.000 variable.py:275(shape)
  1371127    1.909    0.000    3.123    0.000 common.py:1722(_get_dtype_type)
    85683    1.857    0.000   11.168    0.000 nanfunctions.py:1187(nanvar)
    28944    1.820    0.000    1.820    0.000 {method 'astype' of 'numpy.ndarray' objects}
   999649    1.771    0.000    3.821    0.000 variable.py:417(_parse_dimensions)
        6    1.714    0.286   35.320    5.887 dataset.py:2538(reduce)
   199927    1.683    0.000   32.946    0.000 variable.py:1226(reduce)
   371296    1.639    0.000   21.287    0.000 variable.py:40(as_variable)
   399858    1.630    0.000   14.615    0.000 variable.py:1619(__init__)
   142808    1.608    0.000   21.040    0.000 merge.py:174(expand_variable_dicts)
   571240    1.577    0.000    3.475    0.000 numeric.py:2692(seterr)
    85696    1.551    0.000    5.278    0.000 dataset.py:88(calculate_dimensions)
    85685    1.544    0.000   25.092    0.000 alignment.py:171(deep_align)
1771013/1770932    1.529    0.000    3.196    0.000 abc.py:148(__subclasscheck__)
   542731    1.464    0.000    2.521    0.000 base.py:367(_simple_new)
   399857    1.312    0.000    7.747    0.000 base.py:1066(set_names)
   399916    1.275    0.000    1.350    0.000 base.py:701(_get_attributes_dict)
1542459/1542456    1.274    0.000    1.376    0.000 common.py:178(__setattr__)
    85685    1.264    0.000    5.031    0.000 merge.py:254(coerce_pandas_values)
1885175/1799490    1.261    0.000    2.950    0.000 numeric.py:463(asarray)
    57123    1.244    0.000   14.285    0.000 dataarray.py:27(_infer_coords_and_dims)
   571240    1.181    0.000    1.241    0.000 numeric.py:2792(geterr)
  4427056    1.152    0.000    1.152    0.000 variable.py:411(dims)
   456976    1.132    0.000    4.129    0.000 fromnumeric.py:1710(sum)
    28647    1.107    0.000    1.107    0.000 {print}
   114259    1.097    0.000    1.749    0.000 _abcoll.py:412(items)
   399857    1.090    0.000   10.826    0.000 variable.py:1726(to_index)
   799785    1.080    0.000    1.080    0.000 {zip}
   142808    1.065    0.000    8.242    0.000 variable.py:1871(assert_unique_multiindex_level_names)
  1456694    1.055    0.000    1.671    0.000 collections.py:138(iteritems)
   257053    1.021    0.000    1.021    0.000 {numpy.core.multiarray.copyto}
        3    1.016    0.339  223.880   74.627 DataSet.py:1039(_convertToXrDataset)
   399916    0.991    0.000    5.013    0.000 base.py:404(_shallow_copy)
   171366    0.987    0.000    2.643    0.000 nanfunctions.py:36(_replace_nan)
   228496    0.945    0.000   17.822    0.000 dataarray.py:164(__init__)
  1028511    0.940    0.000    1.252    0.000 dtypes.py:84(is_dtype)
  1827943    0.932    0.000    3.086    0.000 indexing.py:1219(shape)
  1856569    0.825    0.000    1.076    0.000 base.py:528(__len__)
   199929    0.813    0.000    6.331    0.000 common.py:140(sizes)
   228522    0.811    0.000    1.491    0.000 contextlib.py:21(__exit__)
    85685    0.810    0.000   15.373    0.000 alignment.py:37(align)
2687305/2687191    0.778    0.000    0.780    0.000 {issubclass}
    57120    0.754    0.000    4.260    0.000 numeric.py:324(equals)
   628463    0.745    0.000    1.052    0.000 common.py:1673(_get_dtype)
    28561    0.718    0.000   17.162    0.001 dataset.py:2691(to_array)
  1199577    0.674    0.000    1.133    0.000 variable.py:120(_maybe_wrap_data)
   399854    0.657    0.000    2.266    0.000 duck_array_ops.py:189(_ignore_warnings_if)
   856907    0.649    0.000    1.274    0.000 base.py:559(values)
    57126    0.642    0.000    3.280    0.000 variable.py:437(_broadcast_indexes)
   285612    0.635    0.000    6.162    0.000 variable.py:1694(copy)
   856911    0.624    0.000    0.624    0.000 {method 'view' of 'numpy.ndarray' objects}
    28562    0.597    0.000   57.571    0.002 DataSet.py:853(_collapseNDtoDataArray)
    85683    0.569    0.000    5.959    0.000 nanfunctions.py:733(nanmean)
    85683    0.551    0.000   11.787    0.000 nanfunctions.py:1328(nanstd)
   199941    0.547    0.000    0.547    0.000 duck_array_ops.py:36(_dask_or_eager_func)
 343665/9    0.544    0.000    1.745    0.194 copy.py:145(deepcopy)
       55    0.523    0.010    4.973    0.090 format.py:1695(_save_chunk)
  1142496    0.519    0.000    0.527    0.000 utils.py:291(__getitem__)
   514104    0.514    0.000    1.295    0.000 variable.py:197(_as_array_or_item)
   114356    0.513    0.000    2.547    0.000 missing.py:119(_isnull_ndarraylike)
    28563    0.510    0.000   39.631    0.001 merge.py:380(merge_core)
   142807    0.510    0.000    9.029    0.000 utils.py:39(safe_cast_to_index)
   142805    0.488    0.000    2.086    0.000 numeric.py:30(__new__)
   571241    0.482    0.000    0.482    0.000 {numpy.core.umath.seterrobj}
   171586    0.475    0.000    0.475    0.000 warnings.py:364(__enter__)
    57122    0.475    0.000    3.146    0.000 coordinates.py:225(_update_coords)
   285721    0.447    0.000    0.845    0.000 dtypes.py:367(is_dtype)
    28566    0.444    0.000    2.759    0.000 dataset.py:773(_construct_dataarray)
   257050    0.437    0.000   20.105    0.000 duck_array_ops.py:51(f)
        5    0.434    0.087    0.603    0.121 internals.py:4473(_stack_arrays)
   199927    0.423    0.000    3.665    0.000 variable.py:709(attrs)
   142806    0.422    0.000    0.422    0.000 cached_ndarray.py:296(__getitem__)
    28562    0.418    0.000    0.607    0.000 variable.py:500(_validate_indexers)
    57122    0.416    0.000   43.846    0.001 coordinates.py:85(update)
   171586    0.415    0.000    0.415    0.000 warnings.py:380(__exit__)
   142807    0.413    0.000    9.758    0.000 indexing.py:1191(__init__)
```

**Comment by wangcj05** [comment]
profile sorted by cumtime:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.010    0.010  302.558  302.558 Driver.py:21(<module>)
        1    0.000    0.000  299.390  299.390 Driver.py:230(raven)
        1    0.000    0.000  299.386  299.386 Simulation.py:730(run)
        2    0.000    0.000  299.385  149.693 Steps.py:299(takeAstep)
        2    0.000    0.000  299.384  149.692 Steps.py:894(_localTakeAstepRun)
      2/1    0.193    0.097  275.668  275.668 DataSet.py:426(load)
        1    0.098    0.098  263.195  263.195 HistorySet.py:113(_fromCSV)
        1    0.047    0.047  224.175  224.175 DataSet.py:1234(_fromDict)
        3    0.107    0.036  223.987   74.662 DataSet.py:264(asDataset)
        3    1.016    0.339  223.880   74.627 DataSet.py:1039(_convertToXrDataset)
    28562    0.597    0.000   57.571    0.002 DataSet.py:853(_collapseNDtoDataArray)
        1   13.851   13.851   55.179   55.179 DataSet.py:919(_convertArrayListToDataset)
    57122    0.148    0.000   49.683    0.001 DataSet.py:1083(<genexpr>)
    28576   49.418    0.002   49.418    0.002 {method 'index' of 'list' objects}
    28562    0.158    0.000   44.726    0.002 DataSet.py:330(getDimensions)
    85684    0.343    0.000   44.568    0.001 DataSet.py:345(<genexpr>)
    57122    0.062    0.000   43.908    0.001 coordinates.py:20(__setitem__)
    57122    0.416    0.000   43.846    0.001 coordinates.py:85(update)
    28562    0.247    0.000   40.938    0.001 dataset.py:319(__init__)
    28562    0.154    0.000   40.203    0.001 dataset.py:370(_set_init_vars_and_dims)
    28562    0.193    0.000   39.828    0.001 merge.py:358(merge_data_and_coords)
    28563    0.510    0.000   39.631    0.001 merge.py:380(merge_core)
    57122    0.375    0.000   38.218    0.001 merge.py:339(merge_coords)
        3    0.275    0.092   36.095   12.032 DataSet.py:1612(_setScalingFactors)
        6    0.000    0.000   35.320    5.887 common.py:50(wrapped_func)
        6    1.714    0.286   35.320    5.887 dataset.py:2538(reduce)
   199927    1.683    0.000   32.946    0.000 variable.py:1226(reduce)
  1770881   11.382    0.000   27.710    0.000 collections.py:50(__init__)
   199927    2.115    0.000   25.143    0.000 duck_array_ops.py:275(f)
    85685    1.544    0.000   25.092    0.000 alignment.py:171(deep_align)
        1    0.000    0.000   23.716   23.716 OutStreamPrint.py:126(addOutput)
        1    0.030    0.030   23.656   23.656 DataSet.py:635(write)
        1    0.036    0.036   21.329   21.329 HistorySet.py:227(_toCSV)
   371296    1.639    0.000   21.287    0.000 variable.py:40(as_variable)
   142808    1.608    0.000   21.040    0.000 merge.py:174(expand_variable_dicts)
   257050    0.437    0.000   20.105    0.000 duck_array_ops.py:51(f)
    28561    0.070    0.000   19.778    0.001 dataarray.py:475(__setitem__)
   142810    2.051    0.000   18.462    0.000 merge.py:110(merge_variables)
   228496    0.945    0.000   17.822    0.000 dataarray.py:164(__init__)
    28561    0.718    0.000   17.162    0.001 dataset.py:2691(to_array)
   999649    2.005    0.000   15.541    0.000 variable.py:242(__init__)
    85685    0.810    0.000   15.373    0.000 alignment.py:37(align)
   399858    1.630    0.000   14.615    0.000 variable.py:1619(__init__)
    57123    1.244    0.000   14.285    0.000 dataarray.py:27(_infer_coords_and_dims)
  1770912    4.509    0.000   12.942    0.000 _abcoll.py:548(update)
    85683    0.551    0.000   11.787    0.000 nanfunctions.py:1328(nanstd)
24507980/24507976    7.442    0.000   11.484    0.000 {isinstance}
    85683    1.857    0.000   11.168    0.000 nanfunctions.py:1187(nanvar)
   399857    1.090    0.000   10.826    0.000 variable.py:1726(to_index)
   142807    0.413    0.000    9.758    0.000 indexing.py:1191(__init__)
   114244    0.234    0.000    9.417    0.000 variable.py:400(to_index_variable)
        2    0.005    0.002    9.244    4.622 DataSet.py:1765(_usePandasWriteCSV)
    85685    0.154    0.000    9.114    0.000 merge.py:298(_get_priority_vars)
   142807    0.510    0.000    9.029    0.000 utils.py:39(safe_cast_to_index)
   199931    0.181    0.000    8.684    0.000 merge.py:41(unique_variable)
    57120    0.131    0.000    8.464    0.000 variable.py:1704(equals)
  1199577    3.208    0.000    8.274    0.000 variable.py:140(as_compatible_data)
   142808    1.065    0.000    8.242    0.000 variable.py:1871(assert_unique_multiindex_level_names)
    57120    0.153    0.000    8.201    0.000 variable.py:1717(_data_equals)
142819/142815    2.455    0.000    8.176    0.000 base.py:152(__new__)
        1    0.296    0.296    8.033    8.033 dataset.py:2336(dropna)
    28561    0.111    0.000    7.864    0.000 DataSet.py:317(constructNDSample)
   399857    1.312    0.000    7.747    0.000 base.py:1066(set_names)
        2    0.250    0.125    7.500    3.750 dataset.py:1355(isel)
   199929    0.327    0.000    6.741    0.000 dataarray.py:536(indexes)
6086771/6001087    3.614    0.000    6.669    0.000 {hasattr}
   228492    0.241    0.000    6.638    0.000 variable.py:1742(level_names)
   199929    0.813    0.000    6.331    0.000 common.py:140(sizes)
    57126    0.387    0.000    6.232    0.000 variable.py:827(isel)
   285612    0.635    0.000    6.162    0.000 variable.py:1694(copy)
   257049    2.078    0.000    6.159    0.000 nanfunctions.py:107(_divide_by_count)
  2856136    2.157    0.000    6.103    0.000 utils.py:425(ndim)
    85683    0.569    0.000    5.959    0.000 nanfunctions.py:733(nanmean)
    57123    0.325    0.000    5.923    0.000 dataarray.py:679(copy)
    57126    0.179    0.000    5.728    0.000 variable.py:607(__getitem__)
  2570663    5.472    0.000    5.472    0.000 collections.py:71(__setitem__)
    85696    1.551    0.000    5.278    0.000 dataset.py:88(calculate_dimensions)
    85685    1.264    0.000    5.031    0.000 merge.py:254(coerce_pandas_values)
        2    0.000    0.000    5.024    2.512 DataSet.py:1555(_readPandasCSV)
        2    0.000    0.000    5.015    2.507 parsers.py:503(parser_f)
```

**Comment by wangcj05** [comment]
It seems the loop over ```DataSet.py:1083``` the index of list is slow.

**Comment by wangcj05** [comment]
@PaulTalbot-INL @alfoa profile of dataset.

**Comment by PaulTalbot-INL** [comment]
That looks more or less like I'd expect it too, without driving in and finding the details. If we can find a way to speed up the collapsing of realizations into the dataset, we can take out most of the issue the current data objects have.

**Comment by alfoa** [comment]
In addition, it looks like that the getDimension methods (dict comprehension)  is quite slow as well... we might want to try to find a better approach...

**Comment by PaulTalbot-INL** [comment]
@alfoa I agree, it looks like this "variables" method is a much better method for finding indexes.

**Comment by wangcj05** [comment]
checklist passed and closed with PR#840
